### ISSUE TITLE
Application Rework

### ISSUE THREAD

**Issue by JayUnt** [RaspberryPints/RaspberryPints#223]
So after many, many talks we have decided that we are going to convert the site to use a php framework.  In keeping with being lightweight, CodeIgnitor is going to be the choice. 

I created this issue for two reasons.  To see if anyone else has a different option they want to throw into the rink before we start. And to see where the other branches are at, development wise.  

If they other branches are done/close to done, they can get into the branch and be re-written.  If they haven't started yet, they should wait until this is done.  And if they are not done yet, well you might be better off waiting until the conversion is done, because it will most likely have to be re-written anyway.

Its going to be an inconvenience, but since the site is still small, this is the best time to make the conversion.  A bunch of stuff might need to be rewritten, but waiting on this will just make the conversion even worse.  

And if anyone knows CodeIgnitor or wants to learn it, let me know, and ill get you involved in the conversion.  The more people the better.  Ive never used it, but i have used many other MVC frameworks, so this should be too hard, right? :)

This issue was initiated by issue #199,

**Comment by mcangeli** [comment]
I'm not familiar with CI, but I'm willing to learn....

**Comment by mcangeli** [comment]
@JayUnt, just curious to why CI? I've sat down and done some searching on CI and some of the other frameworks..

(http://nerdmom.wordpress.com/2012/08/02/comparing-laravel-codeigniter-cakephp/)
(http://heybigname.com/2012/05/06/why-codeigniter-is-dead)

And granted, most of what I've found is a year and a half old at this point, but a lot of what I've read centers around CI's development (or lack there of).

Also, from last july, EllisLabs was looking for a new "owner" of CI, so who knows which direction its going to go in? (most of the comments on that post aren't too favorable...)
http://ellislab.com/blog/entry/ellislab-seeking-new-owner-for-codeigniter

I'm just wondering if other frameworks were looked at and considered (one that comes up as an improved CI is http://laravel.com/ )

Mind you, I come from the early days of programming where if you wanted the code to do something, you coded it yourself. 

I've adapted over time to become familiar with wordpresses way of handling things and now I'm slowly learning the models and managers you've been using here, but for me its still easiest to create the functions as I needed them. 

So it doesn't truly matter to me as I'm learning new either way.

**Comment by kerber** [comment]
@mcangeli, the main consideration was speed, small footprint, and simplicity. I've not looked into it too much yet, but it looks like Laravel has some cool features. Have you seen anything on what kind of performance it gets?

**Comment by mcangeli** [comment]
I haven't, but they look comparable.

http://www.sitepoint.com/goodbye-codeigniter-hello-laravel/
http://www.youtube.com/watch?v=wq8rc6IlQ1g

**Comment by mcangeli** [comment]
Also, I like what I'm seeing about the blade templating in laravel.

**Comment by kerber** [comment]
Very interesting! This hadn't come up in my searching for lightweight frameworks. It would be awesome if we can get all those features and the template engine and not have it be slow. It would be a good idea to do some testing. I'd hate for us to get close to the end of V2 development, only to realize that the stack is slow.

**Comment by mcangeli** [comment]
http://systemsarchitect.net/performance-benchmark-of-popular-php-frameworks/

Looks like its average. Though this was based on an older version.

**Comment by mcangeli** [comment]
@kerber I think this was written for you... http://mulzer.tumblr.com/post/24141993116/12-reasons-you-should-switch-from-codeigniter-to

**Comment by kerber** [comment]
It looks like that benchmark deals pretty exclusively with simple requests per second. This isn't the concen so much as how much extra processing the framework adds per page. Overall, it looks pretty cool.

**Comment by JayUnt** [comment]
I'll definitely take a look at at Laravel.  And if everyone want to work with laravel, I'm down.  I was initially looking between CI and cakePHP, but cakePHP looked too bloated for our pi servers.  I did read about the main developer and how he does it on his free time, but it being open source and widely used, I figured it wasn't an issue.  If laravel is actively developed, that would ease that issue i had with CI.

The other thing was CI supposedly doesn't have great authentication functionality. From what i read,  its there, but confusing.  If laravel has a better/easier one, then thats another point to laravel.

Im down for all opinions on which way to go, since its our project.

**Comment by mcangeli** [comment]
Laravel has auth built in. Its more maintained, it has a built in templating engine (which would allow for easier customization)

**Comment by mcangeli** [comment]
Like I said above either way I'm going to have to learn from 0. 

From what I've read laravel offers more flexibility with the same performance (or close enough that it doesn't matter).

**Comment by thadius856** [comment]
Not enough knowledge to meaningfully contribute here, but...

...are we going to get basic training on how to make releases with this new bundle?

I'm going to be pretty out of the loop before the next 4 weeks are over.

**Comment by kerber** [comment]
We'll need to put together some references and links to docs for the frameworks. I'm going to have to do a lot of learning, so I'll put together some references I pick up along the way. Hopefully this ends up making things sufficiently well organized that development isn't too hard.

**Comment by Boombrewer** [comment]
In either case, is this going to be something that allows those of us that know nothing about PHP coding to actually help out or are we still too far out of the loop?

**Comment by JayUnt** [comment]
You really should know php and mvc architecture to work on a feature. But to me this is a good way for you guys too learn all about it. I'm down with taking a feature and splitting it up into smaller pieces, for others to work on. This would get you guys used to what goes where.  I can get the project setup, do one of the screens/forms, and hand another screen/off to someone else. Then that can just mimic what I did and learn how it's done.  Even do a bit of code review when it's committed. 

The more people that can help, the better we are in the end.

**Comment by JayUnt** [comment]
And @thadius856, from what ive read. the releases should be not very different. We will just need to set some values in the config files.  We can make a checklist on what to do before the day comes.

**Comment by mtnbound** [comment]
I don't mind learning a new framework. The BeerXML is probably 60% done, and the "don't track keg status" is 100% done, but I don't mind redoing it. I've already built the modules and controllers from scratch, so it would be a nice change to have a framework do it.

**Comment by kerber** [comment]
@mtnbound, Are you planning on migrating your changes into the new framework, once it's up? Once the framework is in place and feature porting is happening, I'd love to help with the BeerXML. Love me some data parsing.

**Comment by mtnbound** [comment]
Sure. I was planning on it, but can always use help. I wanted to learn PHP XML parsing, so I picked this feature :)

I hung all the new tables off the Beer table, so the DB redesign shouldn't affect the structure of the BeerXML.

**Comment by kerber** [comment]
Excellent. That sounds really cool!

**Comment by Skemp45** [comment]
I have never personally used a Framework before, so I will have to learn it from the start. Seems like it might slow us down for a while but then hopefully help in the long run, ...... not looking forward to essentially learning a new language lol

**Comment by JayUnt** [comment]
Once you get it, its easy.  It will be good to have some sort of structure, since we have been really just doing whatever we please.  This will make all of our code more consistent.

**Comment by Skemp45** [comment]
For sure

**Comment by mcangeli** [comment]
After playing around with laravel a bit today, I want to play a little devil's advocate....

Currently, the layout and structure of RP is conducive to the DIY, if you don't like something change it. 

Moving to a framework (any framework) makes that harder to do. No longer is it code that I could download and tweak to my desires with little effort (if I already knew php). Now I have to learn the framework to make it work the way I want. It makes it more difficult to customize the code.

In addition, it no longer becomes an easy "download, unzip, copy and enjoy" type setup.

If you look at most of the frameworks, there are at least three directories...

app (where most of the heavy lifting code is), services and public (which is what the people actually see).

Currently, if you want to install RP, you grab the zip. unzip and copy it to /var/www

With a framework, you unzip, and have to move certain directories to certain places (of course, most frameworks recommend that the only folder the public actually has access to is the public folder).

Don't get me wrong, a framework is a good tool that could greatly improve the process, however I think if this was a site like brewprint.com where its a completely hosted solution, then it make sense. 

The last thing we need is to update the code to use a framework, then we get a bazillion support posts where the user has edited something in the code or not put a folder in the right place and now we're supporting it and having to debug a ton of small errors like that.

A better route may be to move the bits of code to a functions.php page that can be called by the index or another page to produce the content. That would still allow us to "publish" some simple calls that could be included anywhere in a page to display the items, but make it easy for the end user to use and install.

**Comment by mcangeli** [comment]
If you want to see what I am talking about, take a look at:
https://github.com/mcangeli/RaspberryPints/tree/Untappd

look at the index.php file and the includes/functions.php file

**Comment by JayUnt** [comment]
I agree that we do want to allow tinkering, but another way that the app could be unfriendly to tinkering is if it gets out of control.  Using a framework will help prevent that, and if someone wants to tinker, they can just look up how to use the framework.  If we don't use a framework and the site gets big, there is no place for the people to go to see why things are working they way they are.

I downloaded Laravel and started to play around with it locally.  My goal was to see what it takes to get the beer listing page to look like it does, but in the framework.  Granted i know php and MVC structure, but its extremely easy.  

Yes, there will have be some configuration changes, like pointing their site to /public/ instead of just /, but i dont think thats a big deal.  I had to install an app to download Laravel and all of its dependencies, but from what i read, once this is in, everyone else can just download it. Since all of the dependencies are on an app level.  If we want to update framework, then i just need to run one command line.

I hear what you are saying, but i still think we should go framwork route.  Or if not framework route, we at least come up with some 

As for your functions.php, i think the mvc framework would clean this up tremendously.  Instead of having a bunch of separate functions, you will have a model, which you just instantiate and then you can call a loop on $untappd->feed.  It separates where you are displaying it, from how its getting the data.  So if we decided to build our own version of untappd, we could just switch where the model gets its data from, and the front end will just work.

**Comment by mcangeli** [comment]
Like I said, I was playing devils advocate. From my background, its easier without the framework. I have no doubt that once I learn the framework I'll be wondering why I never used it before. ;)

and yea, it does seem once its downloaded and thats it.

What I cannot tell though is if the end user would need to download the laravel.phar file or not or if they just need the files for the "app"

**Comment by JayUnt** [comment]
Dont worry, i know you are just trying to spark the conversation to get us to go the right path.  

The laravel.phar is to just get the files you need to make the app and it makes the folder structure too.  Once you get all of that, all of the dependencies are in the /vendors folder of the app and you are good to go.  So after i commit my code, you will be able to download it and just use it.

If we had to get the users to download it and then install Composer and run it all, i wouldn't go the laravel route.

**Comment by mcangeli** [comment]
I agree. :)

In fact, I may sit back and learn the mvc route while you convert the current site to it then start to contribute... :+1:

**Comment by JayUnt** [comment]
Believe you me. You won't be disappointed. :-)

**Comment by paulcanning** [comment]
FYI CodeIgniter needs very minimal setup and you can just plonk the files right into your WWW folder in Apache. Couple of edits in the config file if needed and setting up the DB, but there are no dependencies like Laravel.

Grab the zip, unzip and go. (well, within reason!)

I also believe CI is easier to learn for those here who have never worked with an MVC.

**Comment by JayUnt** [comment]
Paul, in your opinion, is CI still actively developed? Have you used
laravel? From looking around, it seems like a lot of ppl go from CI to
Laravel but not the other way around.

Also, have you implemented authentication with CI? from what I've read,
it's not as simple as it should be.

I just DLed Laravel, and it was very simple to use. I got the site
basically set up and the beers listing page working in a matter of hours.

I think I'm going to do the same with CI to see which seems easier.

**Comment by paulcanning** [comment]
I have used CI in over a handful of projects, all very different.

Whilst I have yet to try Laravel for myself properly, I have looked into it, but I still stick with the fact that CI is super fast for getting a prototype up and running. No dependencies to faff with and like I said, you simply copy the files and the installation should work (with a standard WAMP/XAMP/LAMP setup).

CI does not have an authentication library built in, but there is a great 3rd party library called Ion_auth that I use all the time. Very easy to include and once included with the main project, that's it, contributors won't need to do anything extra.

I do know that people move over to Laravel and I believe it is superior in a few ways, but when you have lots of people on here that have not used a MVC frameowkr before, CI wins for ease of use.

I'm not sure on its current activity, but the current build is rock solid and you still get people making Saprks on http://getsparks.org/ (you wont really need any of these, but it is a CI community, to give you an idea)

I'm confident you could have a beer list up and running in CI in half an hour.

**Comment by JayUnt** [comment]
Cool!  I'll work on it at some point today, and let you know my thoughts.

**Comment by JayUnt** [comment]
I just wanted to give you all an update on my status.  Like i said, i was planning on testing out CI to see how easy it was.  Unfortunately i wasn't able to block out time to do it yet.  Hopefully i can get some time tomorrow to do it.

But so far I do like Laravel.  The whole "having to install an app to download it and all of its dependencies" was a bit annoying, but it was really a one time thing.  And you guys and the users wont need to do it.  The other thing I'm not a huge fan of is the templating system they use, Blade.  It has some good pieces, like being able to setup a master page, and then all of the pages just supply content.  This can be done without Blade, but they make it nice and clean.  Im going to see if there is a way to only use it for pieces, so we don't have to use it for looping and all of that, and instead just use pure php.

Ill let you know more when i get to CI, and then we can all discuss which we want to use.

**Comment by kerber** [comment]
@JayUnt, Sounds great! I wonder - Do we even need a template engine? It's just the tap list that we want to be theme-able, right? It seems like a lot of overkill.

I'm excited to hear your results!

**Comment by paulcanning** [comment]
The basic templating system in CI should be sufficient IMHO

**Comment by JayUnt** [comment]
@paulcanning, When you use CI, do you use a ORM plugin, or do you just use the Active Record that comes with CI?  So far I like CI, but the ORM it doesn't seem as easy to use.  It seems like CI is lightweight because it doesn't include as much functionality, but instead you have to use different plugins to do what is not there.

**Comment by thadius856** [comment]
I feel the same way about GIMP, but greatly appreciate its plugin
functionality.

Only problem is if the plugs get stale version and are no longer supported.
:\

**Comment by paulcanning** [comment]
I use this library to extend Active Record

https://github.com/jamierumbelow/codeigniter-base-model

**Comment by JayUnt** [comment]
Thanks. I'll look into this plugin.

**Comment by JayUnt** [comment]
Ok so this is what I've come up with and just want to run it past everyone.  Between CI and Laravel, I like Laravel better.  My main issues with Laravel are the initial download of compass to get the app, but that is a one time issue. Once it's downloaded no one else needs to worry about it.  The second was the templating system that they ship with. It turns out you don't need to use it, or can use pats of it.  

With CI, it's very lightweight but that's because you need to download plugins, or sparks add they are called. But there are different plugins to do the same thing, so you then have to determine which will bee around longer and which is best.  You don't have to download compass to get the app initially, but it's not a huge game changer.

Laravel might be less lightweight, but their developers are working on these features that we will use, as opposed to random people. 

If anyone has any opinions let me know, but I vote for laravel.  If no objections, I'm going to make a new branch and start the conversion. 

And let me know if you want to be involved in the conversion. I'll pass out tasks so everyone can learn the new structure.

**Comment by thadius856** [comment]
I don't need to be involved (highly time invested in a promotion-required
course at work), but would like to be able to release updates/hotfixes once
we launch.

Do I need to start doing any reading now?

**Comment by JayUnt** [comment]
Releasing and hotfixes will work the same way as before. They just need to
overwrite their files with the new ones, and possibly run db update
scripts.

**Comment by paulcanning** [comment]
CI can be used with zero Sparks.

The two additional libraries I use are MY_Model and Ion_Auth. Both
"installed" with out Sparks.

If you choose Laravel, I guess it's a good excuse for me to learn that :-)

**Comment by mcangeli** [comment]
So @JayUnt, I should sit down and learn laravel?

**Comment by JayUnt** [comment]
Yeah, let's go with Laravel. I'm going to clean up what I have and then
commit it into a new branch tonight or tomorrow.  So you guys can start
reading up on it.

**Comment by mcangeli** [comment]
Alrighty then! I'll buckle down and learn laravel.

**Comment by JayUnt** [comment]
Ok guys, i finally have something to show you.  Im sorry it took so long.

If you want to see what I've been working on you can get the feature-move-to-laravel branch.  I pretty much just setup the admin section, but have done all of my work/tweaking with the beers listing/form.  And i think i have it to where I like it, minus some of the styling (which doesn't match the original just yet)

There is one server based change that you need to make.  There are now a couple of folders in the base of the project.  App and public are the ones we care about.  The rest is for laravel to worry about.  But as a security measure we want to point the webserver to look inside public for the site.  This prevents users from getting to our app folder, which holds the brunt of the work and passwords and stuff. (http://stackoverflow.com/questions/7902119/set-the-apache-htdocs-subfolder-as-git-working-dir

After you get the code, you can go to http://localhost/admin/beers (replacing http://localhost with whatever your server is setup as). You can see the listing of beers, and click on the add or edit buttons to see them in action.

Now to explain what we have here.

Ill start with the M in MVC, Models.  In the app folder we have a, wait for it..... models folder.  And inside there we have a, wait for it... Beer.php file.  This is the beer's model.  if you look inside it, its very simple.  The class name is the name of the model.  Laravel assumes that the table that its linked to is the plural of the model, which it is.  So its automatically hooked up.  You can tell it the table's name if its not the assumed name, but we dont have to.

next we have two propertied, $fillable and $guarded.  These are ways to tell laravel which properties can be set by the app.  You dont need to add these, but i would suggest we do to the programmers know what properties are on the model.

And lastly, beer has a relationship with BeerStyle. so we have a function that we can use as a shortcut to get it each time.  If we had the same column naming convention that they use, all we would need to send in is the first parameter.  but since we dont, we have to give more info.  Not really a big deal though.  We also need to have that BeerStyle model to have this relation.

Before we move onto the controllers, which is really where all of the logic should go for each page. We have to map the urls.  in /app/routes.php you will see a line  `Route::resource('admin/beers', 'BeerController');`  This tells the site that if the url admin/beers/ is hit OR if any subpage is hit, goto the BeerController. You can specify each individual url (which we will want to do at time) or you can use the `Route::resource` to have laravel assume all of the REST functions. (http://laravel.com/docs/controllers#resource-controllers)

Onto the controllers, specifically /app/controllers/BeerController.php.  I followed the structure from this page http://scotch.io/tutorials/simple-laravel-crud-with-resource-controllers, with some modifications, mostly pulling out some common code into private functions.  Each public function represents a page.  So if you go to admin/beer/create.  it looks at the BeerController, becasue we set it up like that in the routes.php.  Then it looks at the subpage, create, and looks for a create function.  If it doesnt find it, itll throw a 404, because it doesnt exist.

This create controller is pretty simple.  Its goal is to show the create form, but there isnt much data it needs off the bat.  The only thing it needs is the data for the dropdown list, which i get in the loadFromViewData() private function.  I put it in a private function because both the create and edit will use it.  And if you look at the edit function, its pretty much the same as the create, but it gets the beer you are going to edit first.  Lastly, all of the data you want to send to the page, you shove into an array, and pass that along to the view, which you are returning.

The line `return View::make('beers.create', $data);` tells the app what view to get.  'beer.create' says to look for /app/views/beers/create.blade.php (or just /app/views/beers/create.php if we arent going to use the templating system.  but im starting to like it more and more as i go).  So open up that folder.

The first line is `@extends('layouts.admin')`.  This says to use the admin.blade.php layout, which you will find in /app/views/layouts/.  Open that sucker up.  This will take care of all of the includes for every admin page.  You setup the headers, footers, sidebars, css and js includes and all of that fun stuff.  

Line 28 `@yield('content')` tells this layout that there is a section that the other views can put whatever they want in.  And this section is called 'content'.  Now if you go back to the create.blade.php, you will see on line 3 `@section('content')`.  This is the page saying, put whatever is between this and the first @stop, into that part of the admin layout.

The rest of this create is really just setting up the title, and form, and including the form inputs.  I put the inputs in another page, so the create and edit can share them.  The form codes you can read about here http://laravel.com/docs/html.

The last thing i want to mention, is from line 4 `{{{ Lang::get('common.addBeer') }}}`.  Instead of putting text all over the site, i suggest we use the Lang.get() function.  This will allow us to do localization down the road.  So if at any point we want to allow the users to see the site in Spanish, we dont have to do conditionals everywhere, we just have to tell Laravel to use the spanish version.  these files are located in /app/lang/.  We are defaulted to english, so it goes into the en folder. then using 'common.addBeer' we specified the common.php file and the addBeer resource.  if you look inside common.php you will see its just an array.  the first item in the array  'addBeer' and the value of that item is the string that will be shown.

So if we wanted to add spanish support, all we would have to do is add an es folder, and then copy this common.php file into there. and then change all of the values to their spanish version. And whamo. 

Hopefully this was helpful to all (or some) of you.  Im going to work on the user authentication next, and then move onto the main tap listing page.  

If you have any questions or want to work on a section let me know.

**Comment by JayUnt** [comment]
Oh and one last thing.  I mentioned that i am starting to really like the templating system, Blade.  It has a ton of helper functions for creating the urls for links, and creating the form elements.  I didnt want to use it at first, because i thought it would be too much for you guys at once, but i suggest we stick with it.  I really like it for that master page feature.  It really makes the actual views clean, without cluttering it with a ton of includes.  Just that one line at the top.

**Comment by JayUnt** [comment]
I added the user authentication.  One thing to note, we were using md5 as the password encryption.  Supposedly its very old so laravel doesn't like it.  The login is setup to first look at the more secure encryption, if it fails, it then tried the md5 encryption.  if the md5 succeeds, we then replace the user's password with the more secure version and log them in.  Its all automatic and the user doesn't see anything happen.

**Comment by mcangeli** [comment]
Nice!

I still have to download and look at the stuff you were working on. I received the adaptor on Saturday that will allow me to hook up an old laptop screen as an HDMI monitor so I've been busy working on building a case for that....

**Comment by kerber** [comment]
I've not had any chance to look through the code, but your description sounds really great! I'm really looking forward to digging in!

**Comment by JayUnt** [comment]
One last thing that you will have to do is run the db update script.  I based my code off of the updated db that i made for @kerber.  So the main tap listing's db view is changed, to work with the new batches table.

**Comment by mcangeli** [comment]
@JayUnt I changed apache to look at /home/pi/raspberrypints/public as the directory for the site.
I exported and dropped my previous DB
I downloaded the zip of the feature-laravel that you created and unzipped to my raspberrypints directory.

if I go to http://rp.skypainterbrewing.com all I get now is a white page.

I can go to http://rp.skypainterbrewing.com/mark.php and it pulls up the phpinfo page I created in the public folder so I know its mapped correctly...

any ideas? everything is 777 in the raspberrypints directory.

**Comment by JayUnt** [comment]
What happens when you go to  http://rp.skypainterbrewing.com/admin?  You
should get a plain looking login screen and can use your existing login
credentials.   i tried it but i get a 404, but i dont know if your site is
external facing.

I'm assuming you did this, but did you put all of the code into
/home/pi/raspberrypints/
and just pointed apache to look at the public folder within there?

On Mon, Mar 17, 2014 at 3:21 PM, mcangeli notifications@github.com wrote:

> @JayUnt https://github.com/JayUnt I changed apache to look at
> /home/pi/raspberrypints/public as the directory for the site.
> I exported and dropped my previous DB
> I downloaded the zip of the feature-laravel that you created and unzipped
> to my raspberrypints directory.
> 
> if I go to http://rp.skypainterbrewing.com all I get now is a white page.
> 
> I can go to http://rp.skypainterbrewing.com/mark.php and it pulls up the
> phpinfo page I created in the public folder so I know its mapped
> correctly...
> 
> any ideas? everything is 777 in the raspberrypints directory.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/RaspberryPints/RaspberryPints/issues/223#issuecomment-37858710
> .

## 

Jason Unterman

**Comment by mcangeli** [comment]
Also, on line 214 in the feature-flow_monitor.sql the R in srmRgb needs to be capitalized on both 214 and 215 otherwise it errors.

**Comment by mcangeli** [comment]
Actually, everything at the end of the feature-flow_monitor.sql script.

If its two parts (kegType) the second word needs to be capitalized or it doesn't like the tables as in the schema.sql file.

Its facing the world. I think its an issue with apache and htaccess (I had this problem before on another machine when trying to setup laravel originally... but I can't remember what fixed it).

Also, I had to go in to the config page for the database and change my username and password.

**Comment by mcangeli** [comment]
And yes.

```
pi@raspberrypi ~/raspberrypints $ ls
app      bootstrap      CONTRIBUTING.md  public                           raspberry-pints.sublime-workspace  server.php
artisan  composer.json  phpunit.xml      raspberry-pints.sublime-project  README.md                          sql
pi@raspberrypi ~/raspberrypints $
```

```
 DocumentRoot /home/pi/raspberrypints/public
```

**Comment by JayUnt** [comment]
Thanks, I updated the sql script and committed.  

Im looking to see if i can find a solution to your problem. I'm developing on windows and don't have an issue, and i know others have done it this way. just a matter of finding what to put where.

...

And I just went to your site and i guess you got that part working :)

I still think you have a .htaccess issue because http://rp.skypainterbrewing.com/admin still gets a 404 but http://rp.skypainterbrewing.com/index.php/admin gets that laravel issue.

**Comment by mcangeli** [comment]
no, I removed the directory.

I downloaded the laravel phar file and I've created a "new" project.

Going to unzip and copy the rpints stuff over now.

**Comment by mcangeli** [comment]
Its weird. If I create a new laravel project in the folder, it works without an issue.

Downloading and unzipping the rpints stuff returns the blank page. Even after I set the database config correct in the database.php file.

**Comment by JayUnt** [comment]
let me try moving my site to a different location and see if i works

**Comment by JayUnt** [comment]
But now that I think of it, i already did this.  I moved it from one folder to another before and it worked.

**Comment by mcangeli** [comment]
All on a windows machine though right?

**Comment by JayUnt** [comment]
that is correct.  I am going to be ordering up a pi real soon, so ill be able to check on there when i do. But for now im only doing windows.  ill look into a rpi vm for windows, so i can test it like prod would see it.

**Comment by mcangeli** [comment]
Ok, well part of the problem is that apache2 on rpi does not have mod_rewrite enabled by default.

so...

```
sudo a2enmod rewrite

sudo /etc/init.d/apache2 restart
```

That fixed the problem with it not rewriting the urls correctly.

But it still returns a blank page.

I may download a WAMP stack and try it on my laptop later if I feel like it. ;)

**Comment by mcangeli** [comment]
Bah.

@JayUnt its working now. 

I had gummed up the permissions and ownership some how.

**Comment by mcangeli** [comment]
Also, it looks like your tables on your localhost are all lowercase.

There are issues with it not being able to find tables due to the upper and lower case of the letters.

**Comment by JayUnt** [comment]
i wonder if that is a config thing.  i dont have any issue with different capitalization.  But i can go through my stuff and make sure they are all the same as the initial db script.  Even when i look at the table names in MySQL Workbench, they are all lowercase.

**Comment by JayUnt** [comment]
I guess we will have to play around with it and write a tutorial, in case we need to have the clients make these changes.

**Comment by mcangeli** [comment]
Yea. I also downloaded and ran composer install in the folder where I put the rpints code so I don't know if that had anything to do with it or not.

Luckily I don't mind wiping everything and starting over if I have to to test... :)

**Comment by JayUnt** [comment]
Yeah.  If it turns out you need to dl an run composer for it to work in linux, we might have to change our plan.  I dont want to have the clients doing all of that.  Im thinking there is a way around that though.

**Comment by JayUnt** [comment]
thanks for you help with this.

**Comment by mcangeli** [comment]
@JayUnt I tried it on a totally different linux machine this time.

(one that I had downloaded laravel on before and created an app following the tutorial so I know it worked)

This is what I had to do:

Download, unzip and move the Rpints code to the folder I want it in. Run schema, then the two db updates to get my database up to date. 

Edit app/config/database.php to have my mysql login information

Download the composer:

```
$ curl -sS https://getcomposer.org/installer | php
```

the once that downloaded in my new rpints directory where the app, public, and composer.json file are run:

```
php composer.phar install
```

once there it comes up, but it does not like the naming of the vwgetactivetaps table. I assume because in the database its vwGetActiveTaps and its looking for vwgetactivetaps.

http://skypainterbrewing.com/rp/public/

**Comment by JayUnt** [comment]
I updated the view's name and committed it yesterday.

Im glad it worked but there are two things i dont like still.  

1) you had to download composer. Im hoping there is a way that we wouldnt need the client to download it

2) your site should just be http://skypainterbrewing.com/rp/ without the public/.  This is to prevent people from getting to the other folders/files.  if i go to http://skypainterbrewing.com/rp/app/, i can see a listing of all of your files.  Granted, there is code at the top to prevent me from seeing the content, but im sure a person who knows what they are doing, would be able to get them.  That is why the server is supposed to point to the /public/ folder.

Now that you have this working, can you try again to point it to the public folder?

**Comment by mcangeli** [comment]
yea, I dropped this on my blogs server instead of the pi, thats why its in a sub folder and is setup the way that its setup.

I've changed the site for the time being to point to that folder so you should be able to go to http://skypainterbrewing.com 

it still has the issue with the view and this is from the latest zip I grabbed 20 minutes ago.

**Comment by JayUnt** [comment]
I havent fully finished the home page yet.  but you can go to http://skypainterbrewing.com/login and login.  Im hoping to get that page mostly working today.

ill take a look to see if i missed a spot in the sql file.

**Comment by mcangeli** [comment]
Yea, It logged me in.

It should be possible to create a bash script that the user can run that would download composer and run the stuff. Even set everything up in apache the way it needs to be.

**Comment by mcangeli** [comment]
Ok, I think I found the issue.

You need ALL of the folders in the root folder and not just the app, bootstrap, public and sql folder.

There are files in the vendor folder that are needed and preventing the code from working right when they aren't there.

thats why its not working when I download the zip until I run the composer install, because thats when the dependencies that are housed in the vendor folder are downloaded.

**Comment by JayUnt** [comment]
ahh, i didnt realize you moved folders around.  yeah, that would do it.  the structure should stay the same, just pointing apache into the public folder.

**Comment by mcangeli** [comment]
Your install here doesn't include the vendor folder. Thats whats missing in the download. if that gets added to the repo here it should work for people when they download it.

**Comment by JayUnt** [comment]
ahh good catch.  it looks like laravel has a .gitignore file that tells git to ignore that folder. ill remove it and commit.

**Comment by JayUnt** [comment]
i removed that line from .gitignore, but now my gui keeps freezing.  ill try to use the cmd in a bit.  maybe its too many files?

**Comment by JayUnt** [comment]
Ok it looks like i was able to add it using the git command line.  So hopefully that fixes it.

I see that these vendor folders also have .gitignore, but i think that will be fine.  its to prevent the vendors' vendors from coming through.  But there is really only one way to find out, and test.

**Comment by mcangeli** [comment]
That works.

I removed the stuff I downloaded this morning and redownloaded the zip for the laravel move.
Unzipped it and moved it to the directory I wanted and after a chmod was able to hit the login page and login to the admin section.

**Comment by JayUnt** [comment]
nice!

I've worked a bit on the homepage.  So you should be able to see the listing "mostly" like the original site.

**Comment by mcangeli** [comment]
Yes. That worked. (except this is an empty DB so there isn't anything in the list...)

**Comment by JayUnt** [comment]
Updated login style to match existing. Added kegs listing/form.

**Comment by mcangeli** [comment]
So.... the only thing missing at this point is the tap list page....

**Comment by JayUnt** [comment]
That and all of the personalization stuff. Brewery logo upload, show/hide
columns, profile info, forgot password.

**Comment by mcangeli** [comment]
So.... after much yelling and screaming and arguing with laravel, it seems I may have gotten the brewery feed to work with the new layout. My goodness was it painful. But I suspect that I won't have as much of an issue getting the img and ratings.

**Comment by JayUnt** [comment]
Awesome!  Let me know when you commit it and ill take a look.

**Comment by kerber** [comment]
Excellent!

**Comment by kerber** [comment]
@mcangeli, do you want to go ahead and create an issue for the new Untapped integration feature? We'll use it to keep track of any conversation around it.

**Comment by mcangeli** [comment]
I've issued a pull request for the code I've gotten.

I still want to go through and comment the code so you can understand what I've done, but I think its pretty clean.

The nice thing is that the untappd model should allow me to add the functions for the rating and the beer img now that I know what I'm doing and how to call the variables.

**Comment by JayUnt** [comment]
I made some comments on your code.  Just want to keep with the MVC structure.  Let me know if you have any questions on my comments.  Its idea is very similar to BeerController@index.

**Comment by mcangeli** [comment]
@JayUnt can you post or send me a copy of the data dump.sql for the current table structure?

I started with a new database so there's no data in it (when I try to run the sql updates on my existing rpints db it errors out every time) and I want to play around with the new table structure and taps/batches.

**Comment by mcangeli** [comment]
Also, in the tap.blade.php file in the home view, the math to figure/display what is left needs to be changed....

in stead of 

```
number_format((($beer->startLiter - $beer->remainAmount) * 128));
```

it should be 

```
number_format((($beer->startLiter - $beer->remainAmount)/0.0296));
```

Same for 

```
number_format(($beer->remainAmount) * 128);
```

```
number_format(($beer->remainAmount)/0.0296);
```

Unless you're doing the math somewhere else. But that will convert the 18.93 liters to 640oz.

**Comment by mcangeli** [comment]
Also in tap.blade.php in the views/home folder...

```
<h1><?php echo $beer->beername; ?></h1>
```

Needs to be:

```
<h1><?php echo $beer->name; ?></h1>
```

For the beer name to display properly.

**Comment by JayUnt** [comment]
I committed a backup into the sql folder.  

I actually want to move that number format stuff into the model, since the view shouldnt really be dealing with that stuff. But when I do, ill change the number.

And ill make that beer->beername change too.  Thanks.

I am currently working on the new batches section.  For some reason its giving me trouble, even though its very similar to the other screens.

**Comment by JayUnt** [comment]
I just wanted to let you guys know that i found a cool open icon library that i am using for the tap ajax loading icon.  Very easy to use.  So if you need an icon for anything, look at this place to see if they have the one you want first.

https://github.com/jxnblk/geomicons-open

**Comment by louking** [comment]
I'm just hoping the upgrade path is easy for all this...

**Comment by kerber** [comment]
@louking: It should be. Since you're reading the low level developer discussions, I admit it does sound rather complex. Most of it is just us setting it up initially, though. As the install is mostly just "copy files to web dir and set up apache", those complexities aren't transferred to the end user.

**Comment by Skemp45** [comment]
Closing as this is not going to happen any longer. sticking with old code

**Comment by paulcanning** [comment]
Really? After 5 months of deciding to use Laravel and doing actual development, its being scrapped for procedural PHP? Shame :(

**Comment by Skemp45** [comment]
Here is the think, it has not taken 5 months to make the move. It took 3 weeks at the most lets even give it 1 month and then 4 months of noone doing anything. So with that its back to the old code where everyone was helping out.

**Comment by mcangeli** [comment]
The move to a framework like laravel was debated for a while. We went back
and forth on it. It had been decided to move to it to help development move
along and make things easier for the devs.

And as far as maintaining the base code it's great.

The problem lies in two things... The first, there isn't a good way to
handle themeing or customization. It takes too many steps with laravel to
make it a unique install. Looking at the post on HBT,  people have changed
the layouts beyond simply changing the background as they need to in order
for it to fit their monitor or other setup. The laravel move could work if
we developed a good, simple theme engine.

Second is the integration of other code or plugins. Having worked with
@jayunt and @kerber on the laravel move and having developed the untappd
integration I can say it's a pain in the ass at the moment. A simple plug
in structure would need to be developed and in place as well as documented
for people to use.

I don't think that the move to laravel should be totally dropped. I do
think that for V2, we should evaluate things and perhaps roll out basic
flow meters (the code works. I've been using it for a while now) and work
on making the laravel branch more complete with the themeing and plugins.

We had said that v2 would be flow metering. The masses might rebel if it's
not.

Here is the think, it has not taken 5 months to make the move. It took 3
weeks at the most lets even give it 1 month and then 4 months of noone
doing anything. So with that its back to the old code where everyone was
helping out.
